### PR TITLE
8314512: IGV: clean up hierarchical layout code

### PR DESCRIPTION
**Pull request by tobiasholenstein**
This refactoring enhances layer and node management by encapsulating computations within LayoutGraph, LayoutLayer, and LayoutNode, improving modularity and delegation of responsibilities for handling dummy nodes and self-edges to dedicated utility classes. Code duplication was reduced by consolidating common logic, such as edge reversal, into reusable methods, and optimizing processes like updating node positions and crossings. Additionally, method names were updated to better reflect their functionality,

### LayoutGraph
The LayoutGraph class is responsible for organizing and arranging a graph's nodes and edges for visual display. It takes a collection of nodes (Vertex) and connections between them (Link) and structures them into layers, creating a hierarchical layout. The class handles complexities like edges that span multiple layers by inserting temporary "dummy" nodes to maintain a clear hierarchy. This organization helps ensure that when the graph is displayed, it is easy to understand and visually coherent, making the relationships between nodes clear and straightforward.

### LayoutLayer
The LayoutLayer class represents a single horizontal layer in a hierarchical graph layout. It holds a list of nodes (LayoutNode) that are all on the same vertical level. This class provides simple methods to manage these nodes: you can add nodes to the layer, calculate the maximum height needed to fit all nodes, center the nodes vertically within the layer, and set their horizontal positions with proper spacing. In essence, LayoutLayer helps organize nodes neatly in a graph, making it easier to display the graph clearly and understand the relationships between nodes.

### LayoutNode
The LayoutNode class represents a node in a hierarchical graph layout. It can be either an actual node from the original graph or a temporary "dummy" node added during the layout process to handle complex edge connections. This class stores important layout information like the node's position (x and y coordinates), size (width and height), layer level, and connections to other nodes through incoming and outgoing edges. It provides methods to calculate optimal positions, manage margins, and handle reversed edges, all aimed at arranging the nodes neatly in layers to create a clear and visually organized graph display.

### LayoutEdge
The LayoutEdge class represents a connection between two nodes (LayoutNode) in a hierarchical graph layout. It stores information about the starting node (from), the ending node (to), and the positions where the edge connects to these nodes. It also keeps track of whether the edge has been reversed, which is useful for handling edges that go against the main flow in hierarchical layouts (like loops or back edges). This class ensures that edges are drawn correctly between nodes, helping to create clear and understandable visualizations of the graph.

`LayoutGraph` with `LayoutLayers`
LayoutGraph

`LayoutLayer` with `LayoutNodes` 
LayoutLayer

### Keeping edges straight until they leave the LayoutLayer
`before`
old

`now`
new

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314512](https://bugs.openjdk.org/browse/JDK-8314512): IGV: clean up hierarchical layout code (**Enhancement** - P4)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) Review applies to [70724601](https://git.openjdk.org/jdk/pull/22402/files/70724601abb19b5e30b770b280b48146155901e5)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22402/head:pull/22402` \
`$ git checkout pull/22402`

Update a local copy of the PR: \
`$ git checkout pull/22402` \
`$ git pull https://git.openjdk.org/jdk.git pull/22402/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22402`

View PR using the GUI difftool: \
`$ git pr show -t 22402`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22402.diff">https://git.openjdk.org/jdk/pull/22402.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22402#issuecomment-2504437703)
</details>

### CODE REVIEW

**Review by robcasloz — COMMENTED** [review]
Hi Toby, thanks for doing this! I realize this changeset is not a pure clean up but actually affects the layouts. Could you summarize the proposed layout edges? (condense output edges, extra vertical margin for nodes with input/output ports, changes to the crossing reduction heuristic, etc).

**Comment by tobiasholenstein** [comment]
> I found a couple of issues in the CFG view by playing around manually:
> 
> 1. I get the following assertion error when opening the CFG view of the "Optimized finished" graph in [this file](https://github.com/user-attachments/files/17931456/foo.zip) and enabling the "Simplify graph" filter:
> 
> Screenshot from 2024-11-27 09-33-33
> 
> 2. With this change, node labels in the CFG view are center-aligned (see right part of screenshot below), instead of left-aligned as expected (see left part of screenshot below):
> 
> Screenshot from 2024-11-27 09-44-46

Thanks for pointing those two things out. They should be fixed now!

**Comment by robcasloz** [comment]
> Thanks for pointing those two things out. They should be fixed now!

Thanks for fixing these!

**Review by robcasloz — COMMENTED** [review]
Thanks for mitigating the performance regression, Toby! I benchmarked a version with your latest Java code changes ("layout-after-opts") and the version with the latest Java code changes and assertions disabled ("layout-after-opts-no-asserts"), see  [results.ods](https://github.com/user-attachments/files/17939304/results.ods). In summary, the Java code changes mitigate the regression to some extent (from the original 1.7x mean slowdown down to 1.3x), and additionally disabling the assertions flips the sign of the result and brings a mean 1.5x **speedup** over the original layout time. That's great, I wasn't aware that enabling assertions on IGV was so expensive!

I have also tested IGV, both manually and automatically by re-enabling assertions and laying out thousands of graphs in the sea of nodes, clustered sea of nodes, and control-flow graph views, with different combinations of filters enabled. I did not find any issue.

I plan to review the actual code changes tomorrow.

**Comment by tobiasholenstein** [comment]
> Perhaps

I think I removed the performance bottlenecks. (it was not the layouting).
- I set setToolTipText(“<HTML>” + generateToolTipText(this.connections) + “</HTML>“); only when needed for `LineWidget`
- I add all created `LineWidgets` as a batch with `connectionLayer.addChildren(newWidgets);`
- This left me with a last bottleneck where 90% of runtime for a large graph was spend in `connectionLayer.addChildren` -> `Widget.addChild` -> `Widget.setConstraint`:
```java
public final void setChildConstraint(Widget child, Object constraint) {
        assert this.children.contains(child); // we spend 90% of runtime on this assert

        if (constraint == null) {
            if (this.constraints != null) {
                this.constraints.remove(child);
            }

        } else {
            if (this.constraints == null) {
                this.constraints = new HashMap();
            }

            this.constraints.put(child, constraint);
        }
    }
```
We end up spending almost all runtime for large graphs on this assert that is part of the Netbeans library code. It has never failed. So the best thing is to run Java with asserts disabled, except for when we debug: this can be done be running `mvn` with -Dnetbeans.run.params="-J-da"
I have modified the igv.sh with this command.

**Comment by chhagedorn** [comment]
Drive-by comment: You could update the class comments for the four classes described above in the PR description with the actual PR descriptions which are more detailed than the current class comments found in the code.

File: src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalLayoutManager.java [hunk 1]
```diff
@@ -23,1700 +23,723 @@
...
+        return graph.getAllNodes();
+    }
 
-        private int pointCount;
+    public static class ReverseEdges {
 
-        @Override
-        protected void run() {
+        static public void apply(LayoutGraph graph) {
+            removeSelfEdges(graph);
```
Comment (robcasloz) [inline src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalLayoutManager.java:75]: The proposed code removes self-edges unconditionally. This is OK for the sea-of-nodes layout, but for the CFG layout we do need to draw self-edges (think about single basic block loops). Here is an (artificially edited) example of how the new algorithm misses drawing a self-edge for B7 (left is current IGV, right is IGV with your proposed changes):

Screenshot from 2024-11-28 10-09-09

Here is the artificially edited graph file that illustrates the issue: [self-edges.zip](https://github.com/user-attachments/files/17945729/self-edges.zip)

**Review by robcasloz — COMMENTED** [review]
Before this change, when the layout quality in the "Stable sea of nodes" view had degraded too much, one could switch to the regular "Sea of nodes" view and back to "Stable sea of nodes", to start again from the regular layout. This is not possible anymore with these proposed changes, which severely limits the usability of the "Stable sea of nodes" view in my opinion. Is this something that can be easily fixed?

**Review by robcasloz — COMMENTED** [review]
In the "sea of nodes" view, clicking on an edge used to select its connected nodes, but not after this change.

**Comment by robcasloz** [comment]
> In the "sea of nodes" view, clicking on an edge used to select its connected nodes, but not after this change.

Here is a patch on top of this PR that re-introduces the missing functionality: https://github.com/openjdk/jdk/commit/389ab05ed5505930fbcda7c864316567f3e0ff08. Fee free to incorporate it into this PR.

File: src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalLayoutManager.java (hunk 1 above)
Comment (tobiasholenstein) [reply src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalLayoutManager.java]: should be fixed now

**Review by robcasloz — APPROVED** [review]
Thanks for addressing my comments Toby, looks good! I ran stress testing again and could not find any issue. I have **not** re-run performance testing on the latest changes but I don't think it should be necessary since they are restricted to the CFG view.

**Comment by tobiasholenstein** [comment]
> Drive-by comment: You could update the class comments for the four classes described above in the PR description with the actual PR descriptions which are more detailed than the current class comments found in the code.

Good idea. done

**Review by robcasloz — APPROVED** [review]

**Review by chhagedorn — APPROVED** [review]
Looks good now and is working as expected! Thanks for including the last updates.

**Comment by tobiasholenstein** [comment]
Thanks for the intensive testing and reviewing @robcasloz and @chhagedorn 

/integrate
